test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/fields/GenericSubObjectFieldDataGenerator.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.logsdb.datageneration.fields;
+
+import org.elasticsearch.core.CheckedConsumer;
+import org.elasticsearch.logsdb.datageneration.FieldDataGenerator;
+import org.elasticsearch.logsdb.datageneration.FieldType;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Generic generator for any type of object field (e.g. "object", "nested").
+ */
+public class GenericSubObjectFieldDataGenerator {
+    private final Context context;
+
+    private final List<ChildField> childFields;
+
+    public GenericSubObjectFieldDataGenerator(Context context) {
+        this.context = context;
+
+        childFields = new ArrayList<>();
+        generateChildFields();
+    }
+
+    public CheckedConsumer<XContentBuilder, IOException> mappingWriter(
+        CheckedConsumer<XContentBuilder, IOException> customMappingParameters
+    ) {
+        return b -> {
+            b.startObject();
+            customMappingParameters.accept(b);
+
+            b.startObject("properties");
+            for (var childField : childFields) {
+                b.field(childField.fieldName);
+                childField.generator.mappingWriter().accept(b);
+            }
+            b.endObject();
+
+            b.endObject();
+        };
+    }
+
+    public CheckedConsumer<XContentBuilder, IOException> fieldValueGenerator() {
+        return b -> {
+            b.startObject();
+
+            for (var childField : childFields) {
+                b.field(childField.fieldName);
+                childField.generator.fieldValueGenerator().accept(b);
+            }
+
+            b.endObject();
+        };
+    }
+
+    private void generateChildFields() {
+        var existingFields = new HashSet<String>();
+        // no child fields is legal
+        var childFieldsCount = context.specification().arbitrary().childFieldCount(0, context.specification().maxFieldCountPerLevel());
+
+        for (int i = 0; i < childFieldsCount; i++) {
+            var fieldName = generateFieldName(existingFields);
+
+            if (context.shouldAddObjectField()) {
```

Comment:
Should we give an equal probability of 50% to generating object and nested fields? It currently seems like we prioritize shouldAddObjectField.

Reply:
Yes, this is just based on my impression that `nested` is less used in reality. I'll do 5% each.